### PR TITLE
add copy instance and replace instance command

### PR DESCRIPTION
**Pull request by fuqcool**
* Add `copy instance` command
`copy instance` takes an instance name and starts a new instance with exactly the same settings as the given one.
* Add `replace instance` command
`replace instance` takes an old instance name and new instance name, attach the new instance to the same load balancer as the old one, and then detach the old instance, and terminate it.

Limitations:
* Right now replace only works for load balancers
* Instance name has to be unique across the region

### CODE REVIEW

File: src/api/aws_util.js [hunk 1]
```diff
@@ -144,6 +144,26 @@ function _get_account_id() {
...
+					Values: [name]
+				},
+				{
+					Name: 'instance-state-name',
+					Values: ['running']
+				}
+			]
+		})
+		.then(function (data) {
+			return data.Reservations[0].Instances[0];
```
Comment (mgfreshour) [inline src/api/aws_util.js]: Perhaps add a check that there isn't more than one?

File: src/api/instance/copy.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+function get_instance_info(region, instanceId) {
+	return AWSProvider
+		.get_ec2(region)
+		.describeInstancesAsync({
+			InstanceIds: [instanceId]
+		})
+		.then(function (data) {
+			return data.Reservations[0].Instances[0];
```
Comment (mgfreshour) [inline src/api/instance/copy.js]: Is there a reason to use a function to return this instead of just assigning it to a variable in the function that uses it?

File: src/api/instance/copy.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
+'use strict';
+
+const Promise = require('bluebird');
+const AWSProvider = require('../aws_provider');
+const AWSUtil = require('../aws_util');
+const logger = require('../../logger');
+
+module.exports = function (regions, instance_name, rename) {
+	const region = regions[0];
```
Comment (mgfreshour) [inline src/api/instance/copy.js]: Is this supposed to only ever take 1 region or loop over a list of them?  If the former, why take an array, or at least check that pre-condition.

File: src/api/instance/replace.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
+'use strict';
+
+const Promise = require('bluebird');
+
+const AWSProvider = require('../aws_provider');
+const AWSUtil = require('../aws_util');
+const logger = require('../../logger');
+
+module.exports = function (regions, target_name, source_name) {
+	const region = regions[0];
```
Comment (mgfreshour) [inline src/api/instance/replace.js]: Same as the comment on the other function.

File: src/api/aws_util.js (hunk 1 above)
Comment (torra) [inline src/api/aws_util.js]: in fact, i would say we should fail if there is more than one

File: src/api/instance/copy.js (hunk 3 above)
Comment (torra) [inline src/api/instance/copy.js]: good catch - this should support working in many regions at the same time

File: src/api/instance/copy.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+			]
+		})
+		.then(function (data) {
+			const volumes = data.Volumes.map(function (vol) {
+				return {
+					DeviceName: vol.Attachments[0].Device,
+					Ebs: {
+						VolumeType: vol.VolumeType,
+						VolumeSize: vol.Size,
+						DeleteOnTermination: true
```
Comment (torra) [inline src/api/instance/copy.js]: is there a reason to not also copy `DeleteOnTermination` param from old instance?

File: src/api/instance/replace.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+					{InstanceId: instanceId}
+				]
+			})
+			.then(function (data) {
+				const healthy = data.InstanceStates.every(function (state) {
+					return state.State === 'InService';
+				});
+
+				if (!healthy) {
+					if (retry) {
```
Comment (torra) [inline src/api/instance/replace.js]: i dont think `retry` should be based on number of attempts. it should be based on the reason for the health check failing. if it fails because of 'Instance registration still in progress' (not sure if that is the exact wording), then we should retry, but if it fails because 'Instance has failed too many health checks', then we should fail the replace

File: tst/api/instance/copy.test.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+	afterEach(function () {
+		sandbox.restore();
+	});
+
+	it('should copy an existing instance', function () {
+		return instance
+			.copy('us-east-1', 'old-instance', 'new-instance')
+			.then(function (newInstanceId) {
+				expect(newInstanceId).eql('456');
```
Comment (torra) [inline tst/api/instance/copy.test.js]: this test should check that the new instance we are creating has all of the right properties matching the old instance, like userdata, instance type, disks, availability zone, etc.

File: src/api/instance/copy.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+							DeleteOnTermination: vol.Attachments[0].DeleteOnTermination
+						}
+					};
+				});
+
+			return {BlockDeviceMappings: volumes};
+		});
+}
+
+function pick(obj, keys) {
```
Comment (mgfreshour) [inline src/api/instance/copy.js]: This is a fairly general function, perhaps move it to a general location?

File: src/api/aws_util.js [hunk 9]
```diff
@@ -144,6 +144,41 @@ function _get_account_id() {
 	});
 }
 
+function _get_instance_by_name(region, name) {
+	return AWSProvider
+		.get_ec2(region)
+		.describeInstancesAsync({
```
Comment (alecbenson) [inline src/api/aws_util.js]: Do you need to check the next page token for these calls? Otherwise your describe instances call might not return the full list of instances.

File: src/cli/parse_args.js [hunk 10]
```diff
@@ -387,6 +387,22 @@ function parse_args (args) {
...
+						.option('s', {
+							alias:    'source',
+							describe: 'Source instance name'
+						})
+						.option('t', {
+							alias:    'target',
+							describe: 'Target instance name'
+						})
+						.option('i', ingress_rules_opt)
+						.example('nemesys replace instance -s 123 -t 456 -r us-east-1',	'Replaces target instance with src instance')
```
Comment (mgfreshour) [inline src/cli/parse_args.js]: 123 and 456 seem odd names.  I know they can be anything, but if I were to see this example text I'd assume they were some sort of ID numbers.

File: src/cli/parse_args.js [hunk 11]
```diff
@@ -442,6 +458,23 @@ function parse_args (args) {
...
+					_common_args(yargs)
+						.option('i', {
+							alias: 'instance',
+							describe: 'Instance name'
+						})
+						.option('n', {
+							alias: 'rename',
+							describe: 'New instance name'
+						})
+						.example('nemesys copy instance -i 123 -n new-name')
```
Comment (mgfreshour) [inline src/cli/parse_args.js]: Same as above with 123

File: src/api/instance/copy.js [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+	return Promise.all(regions.map(function (region) {
+		return copy(region, instance_name, rename);
+	}));
+};
+
+function copy(region, instance_name, rename) {
+	return AWSUtil
+		.get_instance_by_name(region, instance_name)
+		.then(function (instance) {
+			logger.info(`Copy attributes from ${instance_name}`);
```
Comment (alecbenson) [inline src/api/instance/copy.js]: You should put this log inside the copy_instance_attrs function itself, maybe.

File: src/api/instance/copy.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+			return copy_instance_attrs(region, instance.InstanceId);
+		})
+		.then(function (instance_attrs) {
+			return rename_instance(instance_attrs, rename);
+		})
+		.then(function (params) {
+			logger.info(`Start instance ${rename}`);
```
Comment (alecbenson) [inline src/api/instance/copy.js]: Maybe move this log to inside the "runInstances" function?

File: src/api/instance/copy.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+function wait_until_runnning(region, instanceId) {
+	return AWSProvider
+		.get_ec2(region)
+		.waitForAsync('instanceRunning', {
+			InstanceIds: [instanceId]
+		})
+		.then(function (data) {
+			const instance = data.Reservations[0].Instances[0];
+
+			if (instance.State.Name !== 'running') {
```
Comment (alecbenson) [inline src/api/instance/copy.js]: maybe make this state a `const`? You use it a few times

File: src/api/instance/replace.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+			LoadBalancerName: lbName
+		});
+}
+
+function detach_from_lb(region, lbName, instanceId) {
+	return AWSProvider
+		.get_elb(region)
+		.deregisterInstancesFromLoadBalancerAsync({
+			Instances: [{InstanceId: instanceId}],
+			LoadBalancerName: lbName
```
Comment (alecbenson) [inline src/api/instance/replace.js]: What would happen if the load balancer name is invalid here (and in the above functions)?

File: src/api/instance/replace.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+	}).then(function () {
+		return get_instance_lb(region, target.InstanceId);
+	}).then(function (name) {
+		lbName = name;
+	}).then(function () {
+		logger.info(`Attach ${source_name} to ${lbName}`);
+		return attach_to_lb(region, lbName, source.InstanceId);
+	}).then(function () {
+		logger.info(`Health check for ${source_name}`);
+		return wait_until_healthy(region, lbName, source.InstanceId);
```
Comment (alecbenson) [inline src/api/instance/replace.js]: Should this health check happen before the instance is attached to the load balancer?

Comment (fuqcool) [inline src/api/instance/replace.js]: It is the load balancer who does the health check. So yea, we need to add instance to it first.

File: src/api/instance/replace.js (hunk 15 above)
Comment (fuqcool) [inline src/api/instance/replace.js]: we expect the name to be valid, if it is not, it should throw an exception.

File: src/api/instance/copy.js [hunk 17]
```diff
@@ -0,0 +1,167 @@
+'use strict';
+
+const Promise = require('bluebird');
```
Comment (mlodz) [inline src/api/instance/copy.js]: Is there a reason to use bluebird promises rather than the built in ones?

Comment (fuqcool) [inline src/api/instance/copy.js]: for historic reasons, we've been using `bluebird`, `q` and builtin promise libraries, we still do, i just wanna be consistent with the rest of the library.

File: src/api/instance/copy.js [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+
+function copy_instance_attrs(region, instanceId) {
+	return Promise
+		.all([
+			get_instance_info(region, instanceId),
+			get_instance_userdata(region, instanceId),
+			get_instance_volumes(region, instanceId)
+		])
+		.spread(function () {
+			let args = [].slice.call(arguments);
```
Comment (mlodz) [inline src/api/instance/copy.js]: The logic on these two lines is hard to follow. Maybe add a comment?

File: src/api/aws_util.js (hunk 9 above)
Comment (fuqcool) [inline src/api/aws_util.js]: Thank you for reminding. In this particular case, i think it's a bit over kill to do pagination. If there are more than 400 instances with the same name, we must have some serious problems. :p

File: src/api/instance/copy.js (hunk 8 above)
Comment (mlodz) [inline src/api/instance/copy.js]: There's a lodash function by the same name that seems to do the same thing. Any reason we can't use lodash?

File: src/api/instance/copy.js (hunk 14 above)
Comment (fuqcool) [inline src/api/instance/copy.js]: technically, it's used once :)

File: src/api/instance/copy.js (hunk 8 above)
Comment (fuqcool) [inline src/api/instance/copy.js]: I killed it because it's used only once, and not very necessary.

File: src/api/instance/copy.js (hunk 17 above)
Comment (torra) [inline src/api/instance/copy.js]: i would use bluebird as `BB`, and native promises when possible. bluebird is great for things like `bluebird.all()`, but i would prefer `new Promise()` if all you need is a Promise object

File: src/api/instance/copy.js (hunk 13 above)
Comment (fuqcool) [inline src/api/instance/copy.js]: putting logs inside functions is good, but in the mean time, you lose the flexibility of what to log and when to log. I will keep it as it is and wait for future requirements to verify.

**Comment by fuqcool** [comment]
Thank you all for the comments, appreciated!
